From f05f5496583b877f194b066f90b8730466dc9937 Mon Sep 17 00:00:00 2001
From: Phil Steitz <psteitz@apache.org>
Date: Mon, 13 Dec 2010 00:43:57 +0000
Subject: [PATCH] Added a getUniqueCount() method to Frequency to return the
 number of unique values included in the frequency table.

Reported and patched by Patrick Meyer
JIRA: MATH-414

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/math/branches/MATH_2_X@1044981 13f79535-47bb-0310-9956-ffa450edef68
---
 .../java/org/apache/commons/math/stat/Frequency.java   | 10 ++++++++++
 src/site/xdoc/changes.xml                              |  4 ++++
 .../org/apache/commons/math/stat/FrequencyTest.java    | 10 ++++++++++
 3 files changed, 24 insertions(+)

diff --git a/src/main/java/org/apache/commons/math/stat/Frequency.java b/src/main/java/org/apache/commons/math/stat/Frequency.java
index 84fa4d9b96..e66c1d327a 100644
--- a/src/main/java/org/apache/commons/math/stat/Frequency.java
+++ b/src/main/java/org/apache/commons/math/stat/Frequency.java
@@ -287,6 +287,16 @@ public long getCount(long v) {
     public long getCount(char v) {
         return getCount(Character.valueOf(v));
     }
+    
+    /**
+     * Returns the number of values in the frequency table.
+     * 
+     * @return the number of unique values that have been added to the frequency table.
+     * @see #valuesIterator() 
+     */
+    public int getUniqueCount(){
+        return freqTable.keySet().size();
+    }
 
     //-------------------------------------------------------------
 
diff --git a/src/site/xdoc/changes.xml b/src/site/xdoc/changes.xml
index 2566e14c2a..5308835719 100644
--- a/src/site/xdoc/changes.xml
+++ b/src/site/xdoc/changes.xml
@@ -52,6 +52,10 @@ The <action> type attribute can be add,update,fix,remove.
     If the output is not quite correct, check for invisible trailing spaces!
      -->
     <release version="2.2" date="TBD" description="TBD">
+      <action dev="psteitz" type="update" issue="MATH-448" due-to="Patrick Meyer">
+        Added a getUniqueCount() method to Frequency to return the number of unique
+        values included in the frequency table.
+      </action>
       <action dev="psteitz" type="fix" issue="MATH-414">
         Modified NormalDistributionImpl.cumulativeProbability to return 0 or 1,
         respectively for values more than 40 standard deviations from the mean.
diff --git a/src/test/java/org/apache/commons/math/stat/FrequencyTest.java b/src/test/java/org/apache/commons/math/stat/FrequencyTest.java
index fcd3717469..a29d0ddba1 100644
--- a/src/test/java/org/apache/commons/math/stat/FrequencyTest.java
+++ b/src/test/java/org/apache/commons/math/stat/FrequencyTest.java
@@ -273,5 +273,15 @@ public void testSerial() {
         f.addValue(twoI);
         assertEquals(f, TestUtils.serializeAndRecover(f));
     }
+    
+    public void testGetUniqueCount() {
+        assertEquals(0, f.getUniqueCount());
+        f.addValue(oneL);
+        assertEquals(1, f.getUniqueCount());
+        f.addValue(oneL);
+        assertEquals(1, f.getUniqueCount());
+        f.addValue(twoI);
+        assertEquals(2, f.getUniqueCount());
+    }
 }
 
